File: todoapp/ios/todoapp/src/GoBinding/GoBinding.swift
import Foundation
import SwiftProtobuf
import Binding

final class GoBinding {
    var binding: BindingBindingProtocol?

    private init() {
        open()
    }

    static let shared: GoBinding = {
        let instance = GoBinding()
        return instance
    }()

    func open() {
        let error: ErrorPointer = nil
        let b = BindingNew(GetDatabasePath(), error)
        self.binding = b
        if (error != nil) {
            print(error ?? "no error")
        }
    }

    func executeCommand(_ payload: SwiftProtobuf.Message, timestamp: Date) {
        do {
            let pd = try payload.serializedData()
            let type = self.getCommandType(payload)
            if type == CommandType.unknown {
                let s = try payload.jsonString()
                print("Unknown command type for: " + s)
            }

            let c = Command.with({
                $0.type = self.getCommandType(payload)
                $0.timestamp = Google_Protobuf_Timestamp(date: timestamp)
                $0.payload = pd
            })

            let cd = try c.serializedData()
            try self.binding?.execute(cd)
        } catch {
            print("Unexpected error: \(error).")
        }
    }

    func getCommandType(_ payload: SwiftProtobuf.Message) -> CommandType {
        switch payload {
        case is Command.NewTask:
            return CommandType.newTask
        case is Command.NewProject:
            return CommandType.newProject
        case is Command.RemoveTask:
            return CommandType.removeTask
        case is Command.AddTask:
            return CommandType.addTask
        case is Command.RenameProject:
            return CommandType.renameProject
        case is Command.RenameTask:
            return CommandType.renameTask
        default:
            return CommandType.unknown
        }
    }

    func getViewModel(location: Location) -> Data {
        do {
            let cd = try location.serializedData()
            let data = try binding!.viewModel(cd)
            return data
        } catch {

            print("Unexpected error: \(error).")
        }

        return Data()
    }

    func close() {
        do {
            try binding!.close()
        } catch {
            print("unexpected error: \(error).")
        }
    }
}

func GetDatabasePath() -> String {
    let path = NSSearchPathForDirectoriesInDomains(
            .documentDirectory, .userDomainMask, true
    ).first!

    let fileName = "\(path)/boltdb";

    return fileName
}

File: todoapp/android/todoapp/app/src/main/java/com/example/todolist/service/GoBinding.kt
package com.example.todolist.service

import android.os.Environment
import binding.Binding
import binding.Binding_
import com.example.todolist.model.CommandOuterClass.Command
import com.example.todolist.model.CommandOuterClass.CommandType
import com.example.todolist.model.ViewModelOuterClass.Location
import com.example.todolist.model.ViewModelOuterClass.LocationType
import com.google.protobuf.Message
import java.io.File

class GoBinding {
    private var binding: Binding_

    init {
        val path: String = Environment
            .getExternalStorageDirectory()
            .getAbsolutePath()

        val dbDirPath = "$path/Android/data/com.example.todolist"
        val dbDir = File(dbDirPath)
        dbDir.mkdirs()

        binding = Binding.new_("$dbDirPath/boltdb")
    }

    fun read(payload: Message): ByteArray? {
        return binding.viewModel(payload.toByteArray())
    }

    fun execute(payload: Message) {
        binding.execute(payload.toByteArray())
    }

    fun close() {
        binding.close()
    }
}
